lib/ical/recur_iterator.js

```diff
@@ -844,6 +850,13 @@ class RecurIterator {
       this.days_index = 0;
       do {
         this.increment_year(this.rule.interval);
+        if (this.has_by_data("BYMONTHDAY")) {
+          this.by_data.BYMONTHDAY = this.normalizeByMonthDayRules(
+            this.last.year,
+            this.last.month,
+            this.rule.parts.BYMONTHDAY
+          );
+        }
```

Comment:
This is the important bit (and the same code just below) – go back to the rule as specified and recalculate the valid days for the new month. This deals with negative rules and puts them all in order.